extensions/amp-list/1.0/component.js

```diff
@@ -0,0 +1,222 @@
+
+  const ioOptions = useMemo(() => {
+    const bufferPct = Math.floor(viewportBuffer * 100);
+    return /** @type {IOOptions} */ {
+      rootMargin: `0% 0% ${bufferPct}% 0%`,
+      threshold: 0,
+    };
+  }, [viewportBuffer]);
+
+  const bottomRef = useRef(null);
+  const isBottomNearingViewport = useIsInViewport(bottomRef, ioOptions);
+
+  const {error, hasMore, loadMore, loading, pages, reset} = useInfiniteQuery({
+    fetchPage: async ({pageParam: nextUrl = src}) => {
+      if (!renderable) {
+        return null;
+      }
+      const page = await fetchJson(nextUrl);
+      return page;
+    },
+    getNextPageParam(lastPage) {
+      return getNextUrl(lastPage, loadMoreBookmark);
+    },
+  });
+
+  // Reset when these props change:
+  useEffect(() => {
+    reset();
+  }, [src, renderable, loadMoreBookmark, reset]);
+
+  const shouldLoadMore =
+    renderable &&
+    loadMoreMode === 'auto' &&
+    isBottomNearingViewport &&
+    !loading &&
+    hasMore;
+
+  useEffect(() => {
+    if (shouldLoadMore) {
+      loadMore();
+    }
+  }, [shouldLoadMore, loadMore]);
+
+  // Rendering logic:
+  const list = useMemo(() => {
+    let items = collectItemsFromPages(pages, itemsKey);
+
+    if (maxItems > 0 && items.length > maxItems) {
+      items = items.slice(0, maxItems);
+    }
+
+    return items.map((item, i) => {
+      let renderedItem = itemTemplate(item);
+
+      // The template engine outputs raw HTML:
+      if (
+        renderedItem &&
+        typeof renderedItem === 'object' &&
+        typeof renderedItem.__html === 'string'
+      ) {
+        renderedItem = <span dangerouslySetInnerHTML={renderedItem} />;
+      }
+
+      return augment(renderedItem, {'key': i, 'role': 'listitem'});
+    });
+  }, [pages, itemsKey, maxItems, itemTemplate]);
+
+  const showLoading = loading && (pages.length === 0 || resetOnRefresh);
+  const showResults = list.length !== 0 && !showLoading;
+  const showLoadMore = loadMoreMode === 'manual' && hasMore;
+
+  useImperativeHandle(
+    ref,
+    () =>
+      /** @type {!BentoListDef.BentoListApi} */ ({
+        refresh: reset,
+      }),
+    [reset]
+  );
+
+  return (
+    <ContainWrapper aria-live="polite" {...rest}>
+      <Fragment test-id="contents">
+        {showLoading && loadingTemplate?.()}
+        {showResults && augment(wrapperTemplate(list), {'role': 'list'})}
+        {error && errorTemplate?.(error)}
+        {showLoadMore &&
+          augment(loadMoreTemplate(), {onClick: () => loadMore()})}
+        {loadMoreMode === 'auto' && <span ref={bottomRef} />}
+      </Fragment>
+    </ContainWrapper>
+  );
+}
+
+const BentoList = forwardRef(BentoListWithRef);
+BentoList.displayName = 'List';
+export {BentoList};
+
+/**
+ * Augments the component with properties
+ *
+ * @param {PreactDef.Renderable} component
+ * @param {object} props
+ * @return {!PreactDef.Renderable}
+ */
+function augment(component, props) {
+  if (!Preact.isValidElement(component)) {
```

Comment:
Nit: import as a named `isValidElement`. We use the `Preact.*` mainly for the JSX transformer.

Reply:
Updated.